src/Generator/Passes/DelegatesPass.cs

```diff
@@ -37,177 +20,199 @@ public DelegatesPass()
+            return new QualifiedType(new TypedefType { Declaration = @delegate });
+        }
+
+        private TypedefDecl GetDelegate(Declaration decl, FunctionType newFunctionType, string delegateName)
+        {
+            DeclarationContext namespaceDelegates = GetDeclContextForDelegates(decl.Namespace);
+            var delegateType = new QualifiedType(new PointerType(new QualifiedType(newFunctionType)));
+            var access = decl is Method ? AccessSpecifier.Private : AccessSpecifier.Public;
+            var existingDelegate = allTypedefs.SingleOrDefault(t => t.Name == delegateName);
+            if (existingDelegate != null)
+            {
+                // Code to ensure that if a delegate is used for a virtual as well as something else, it finally ends up as public
+                if (existingDelegate.Access == AccessSpecifier.Private && access == AccessSpecifier.Public)
+                    existingDelegate.Access = access;
+                // Code to check if there is an existing delegate with a different calling convention
+                // Add the new delegate with calling convention appended to it's name
+                if (((FunctionType) existingDelegate.Type.GetPointee()).CallingConvention ==
+                    newFunctionType.CallingConvention)
+                    return existingDelegate;
+                var @delegate = new TypedefDecl
+                {
+                    Access = access,
+                    Name = delegateName + newFunctionType.CallingConvention,
+                    Namespace = namespaceDelegates,
+                    QualifiedType = delegateType,
+                    IsSynthetized = true
+                };
+                allTypedefs.Add(@delegate);
                 return @delegate;
+            }
+            existingDelegate = new TypedefDecl
+                {
+                    Access = access,
+                    Name = delegateName,
+                    Namespace = namespaceDelegates,
+                    QualifiedType = delegateType,
+                    IsSynthetized = true
+                };
+            allTypedefs.Add(existingDelegate);
+            return existingDelegate;
+        }
+
+        private FunctionType GetNewFunctionType(ITypedDecl decl, QualifiedType type)
+        {
+            var functionType = new FunctionType();
+            var method = decl as Method;
+            if (method != null && method.FunctionType == type)
+            {
+                functionType.Parameters.AddRange(
+                    method.GatherInternalParams(Context.ParserOptions.IsItaniumLikeAbi, true));
+                functionType.CallingConvention = method.CallingConvention;
+                functionType.IsDependent = method.IsDependent;
+                functionType.ReturnType = method.ReturnType;
+            }
+            else
+            {
+                var funcTypeParam = (FunctionType) decl.Type.Desugar().GetFinalPointee().Desugar();
```

Comment:
Add a new `FunctionType(FunctionType)` clone constructor like we have for decls and this gets a lot more clean